beacon_node/beacon_chain/src/observed_aggregates.rs

```diff
@@ -94,26 +162,28 @@ pub enum Error {
     },
 }
 
-/// A `HashSet` that contains entries related to some `Slot`.
-struct SlotHashSet {
-    set: HashSet<Hash256>,
+/// A `HashMap` that contains entries related to some `Slot`.
+struct SlotHashSet<I> {
+    /// Contains a vector of disjoint aggregation bitfields/bitvectors keyed to the hash of
+    /// the corresponding data.
```

Comment:
I think we might want to tweak the wording here, and maybe the implementation too.

At the moment we check whether the new message is a subset of any existing one, but we _don't_ check if the existing messages are a subset of the new one. This means that the cache could end up containing unnecessary messages like `[0x0001, 0x0011, 0x0111, 0x1111]` instead of just `[0x1111]`.

If we modified the insert function to check subsets in the reverse direction and _replaced_ the existing entry if `existing.is_subset(new)` then I think we'd get a more minimally sized cache.

The description could be updated to:

> Contains a vector of maximally-sized aggregation bitfields/bitvectors such that no bitfield/bitvector is a subset of any other in the list.

Reply:
Wow this is super smart. Thank you!

Reply:
Implemented in [ad423ae](https://github.com/sigp/lighthouse/pull/3493/commits/ad423ae5c03518264b623050e65aaf44870dcba6)